File: apps/android/app/src/main/java/chat/mural/network/GeminiLiveTransport.kt
package chat.mural.network

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.media.AudioAttributes
import android.media.AudioDeviceInfo
import android.media.AudioFocusRequest
import android.media.AudioFormat
import android.media.AudioManager
import android.media.AudioRecord
import android.media.AudioTrack
import android.media.MediaRecorder
import android.os.Build
import java.nio.ByteBuffer
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicLong
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeout
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.put
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString
import okio.ByteString.Companion.toByteString

class GeminiLiveTransport(
    private val context: Context,
    private val scope: CoroutineScope,
) {
    var onEvent: ((JsonObject) -> Unit)? = null
    var onFailure: ((String) -> Unit)? = null
    var onLevels: ((Double, Double) -> Unit)? = null

    private val generation = AtomicLong(0)
    private val started = AtomicBoolean(false)
    private val muted = AtomicBoolean(false)
    private var webSocket: WebSocket? = null
    private var audioRecord: AudioRecord? = null
    private var audioTrack: AudioTrack? = null
    private var recordThread: Thread? = null
    private var playThread: Thread? = null
    private val audioScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val json = Json { ignoreUnknownKeys = true }
    private var sessionStartTime: Long = 0

    val isStarted: Boolean get() = started.get()
    val isMuted: Boolean get() = muted.get()

    suspend fun connect(
        apiKey: String,
        model: String,
        instructions: String,
        language: String? = null,
    ) {
        val attemptGeneration = generation.incrementAndGet()

        if (context.checkSelfPermission(Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
            throw GeminiTransportException.microphone
        }

        withContext(Dispatchers.IO) {
            configureAudio()

            val wsUrl = "wss://generativelanguage.googleapis.com/ws/google.ai.generativelanguage.v1beta.GenerativeService.BidiGenerateContent?key=$apiKey"

            val client = OkHttpClient.Builder()
                .build()

            val request = Request.Builder()
                .url(wsUrl)
                .build()

            val setupMessage = buildJsonObject {
                put("setup", buildJsonObject {
                    put("model", "models/$model")
                    put("generationConfig", buildJsonObject {
                        put("responseModalities", buildJsonArray { add(JsonPrimitive("AUDIO")) })
                        put("speechConfig", buildJsonObject {
                            put("voiceConfig", buildJsonObject {
                                put("prebuiltVoiceConfig", buildJsonObject {
                                    put("voiceName", "Kore")
                                })
                            })
                        })
                    })
                    put("systemInstruction", buildJsonObject {
                        put("parts", buildJsonArray {
                            add(buildJsonObject { put("text", instructions) })
                        })
                    })
                    put("inputAudioTranscription", buildJsonObject {})
                })
            }

            val deferred = CompletableDeferred<Unit>()

            webSocket = client.newWebSocket(request, object : WebSocketListener() {
                override fun onOpen(webSocket: WebSocket, response: Response) {
                    android.util.Log.d("GeminiLive", "Connected (HTTP ${response.code})")
                    webSocket.send(setupMessage.toString())
                }

                override fun onMessage(webSocket: WebSocket, bytes: okio.ByteString) {
                    val text = bytes.utf8()
                    try {
                        val msg = json.parseToJsonElement(text).jsonObject
                        handleServerMessage(msg, attemptGeneration)
                        if (msg.containsKey("setupComplete")) {
                            deferred.complete(Unit)
                        }
                    } catch (_: Exception) {}
                }

                override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                    webSocket.close(code, reason)
                }

                override fun onMessage(webSocket: WebSocket, text: String) {
                    try {
                        val msg = json.parseToJsonElement(text).jsonObject
                        handleServerMessage(msg, attemptGeneration)
                        if (msg.containsKey("setupComplete")) {
                            deferred.complete(Unit)
                        }
                    } catch (_: Exception) {}
                }

                override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                    if (attemptGeneration == generation.get()) {
                        onFailure?.invoke(t.message ?: "WebSocket connection failed")
                    }
                    deferred.completeExceptionally(t)
                }

                override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                    if (attemptGeneration == generation.get()) {
                        started.set(false)
                        stopAudio()
                    }
                }
            })

            try {
                withTimeout(15_000L) { deferred.await() }
                started.set(true)
                sessionStartTime = System.currentTimeMillis()
                onEvent?.invoke(buildJsonObject {
                    put("type", JsonPrimitive("mural.session.created"))
                    put("session", buildJsonObject { put("id", JsonPrimitive("gemini-${System.currentTimeMillis()}")) })
                })
                onEvent?.invoke(buildJsonObject {
                    put("type", JsonPrimitive("session.started"))
                    put("session", buildJsonObject { put("id", JsonPrimitive("gemini-${System.currentTimeMillis()}")) })
                })
                startAudioStreaming(attemptGeneration)
            } catch (e: Exception) {
                webSocket?.close(1000, "Connection failed")
                throw GeminiTransportException.connection
            }
        }
    }

    private fun handleServerMessage(msg: JsonObject, expectedGeneration: Long) {
        if (generation.get() != expectedGeneration) return

        val serverContent = msg["serverContent"] as? JsonObject ?: return

        val inputTranscription = serverContent["inputTranscription"] as? JsonObject
        if (inputTranscription != null) {
            val text = (inputTranscription["text"] as? JsonPrimitive)?.contentOrNull
            if (!text.isNullOrEmpty()) {
                val elapsed = System.currentTimeMillis() - sessionStartTime
                onEvent?.invoke(buildJsonObject {
                    put("type", JsonPrimitive("session.input_transcript.delta"))
                    put("event_id", JsonPrimitive(java.util.UUID.randomUUID().toString()))
                    put("delta", JsonPrimitive(text))
                    put("start_ms", JsonPrimitive((elapsed - 500).coerceAtLeast(0).toInt()))
                    put("end_ms", JsonPrimitive(elapsed.toInt()))
                })
            }
        }

        val modelTurn = serverContent["modelTurn"] as? JsonObject
        if (modelTurn != null) {
            val parts = modelTurn["parts"] as? kotlinx.serialization.json.JsonArray ?: return
            for (part in parts) {
                val p = part as? JsonObject ?: continue
                val inlineData = p["inlineData"] as? JsonObject
                if (inlineData != null) {
                    val mimeType = (inlineData["mimeType"] as? JsonPrimitive)?.contentOrNull ?: continue
                    val data = (inlineData["data"] as? JsonPrimitive)?.contentOrNull ?: continue
                    if (mimeType.startsWith("audio/")) {
                        val audioData = android.util.Base64.decode(data, android.util.Base64.DEFAULT)
                        playAudioData(audioData)
                    }
                    continue
                }
                val text = (p["text"] as? JsonPrimitive)?.contentOrNull
                if (!text.isNullOrEmpty()) {
                    val elapsed = System.currentTimeMillis() - sessionStartTime
                    onEvent?.invoke(buildJsonObject {
                        put("type", JsonPrimitive("session.output_transcript.delta"))
                        put("event_id", JsonPrimitive(java.util.UUID.randomUUID().toString()))
                        put("delta", JsonPrimitive(text))
                        put("start_ms", JsonPrimitive((elapsed - 500).coerceAtLeast(0).toInt()))
                        put("end_ms", JsonPrimitive(elapsed.toInt()))
                    })
                }
            }
        }

        val turnComplete = serverContent["turnComplete"]?.toString()?.toBooleanStrictOrNull() == true
        if (turnComplete) {
            val elapsed = System.currentTimeMillis() - sessionStartTime
            onEvent?.invoke(buildJsonObject {
                put("type", JsonPrimitive("session.output_transcript.done"))
                put("event_id", JsonPrimitive(java.util.UUID.randomUUID().toString()))
                put("start_ms", JsonPrimitive((elapsed - 1000).coerceAtLeast(0).toInt()))
                put("end_ms", JsonPrimitive(elapsed.toInt()))
            })
            onLevels?.invoke(0.0, 0.0)
        }
    }

    fun sendAudio(pcmData: ByteArray) {
        if (muted.get() || !started.get()) return
        val ws = webSocket ?: return

        // Calculate input audio level (RMS of 16-bit PCM)
        var sum = 0.0
        var i = 0
        while (i < pcmData.size - 1) {
            val sample = (pcmData[i].toInt() and 0xFF or (pcmData[i + 1].toInt() shl 8)).toShort()
            sum += sample.toDouble() * sample.toDouble()
            i += 2
        }
        val rms = kotlin.math.sqrt(sum / (pcmData.size / 2))
        val inputLevel = (rms / Short.MAX_VALUE).coerceIn(0.0, 1.0)
        onLevels?.invoke(inputLevel, 0.0)

        val base64Data = android.util.Base64.encodeToString(pcmData, android.util.Base64.NO_WRAP)
        val msg = buildJsonObject {
            put("realtimeInput", buildJsonObject {
                put("audio", buildJsonObject {
                    put("mimeType", JsonPrimitive("audio/pcm;rate=16000"))
                    put("data", JsonPrimitive(base64Data))
                })
            })
        }
        ws.send(msg.toString())
    }

    fun sendText(text: String) {
        val ws = webSocket ?: return
        val msg = buildJsonObject {
            put("clientContent", buildJsonObject {
                put("turns", buildJsonArray {
                    add(buildJsonObject {
                        put("role", JsonPrimitive("user"))
                        put("parts", buildJsonArray {
                            add(buildJsonObject { put("text", JsonPrimitive(text)) })
                        })
                    })
                })
                put("turnComplete", JsonPrimitive(true))
            })
        }
        ws.send(msg.toString())
    }

    fun setMute(value: Boolean) {
        muted.set(value)
    }

    fun disconnect() {
        generation.incrementAndGet()
        started.set(false)
        stopAudio()
        webSocket?.close(1000, "Client disconnect")
        webSocket = null
    }

    private fun configureAudio() {
        val sampleRate = 16000
        val channelConfig = AudioFormat.CHANNEL_IN_MONO
        val audioFormat = AudioFormat.ENCODING_PCM_16BIT
        val bufferSize = AudioRecord.getMinBufferSize(sampleRate, channelConfig, audioFormat)

        audioRecord = AudioRecord(
            MediaRecorder.AudioSource.VOICE_COMMUNICATION,
            sampleRate,
            channelConfig,
            audioFormat,
            bufferSize
        )

        val outBufferSize = AudioTrack.getMinBufferSize(
            sampleRate,
            AudioFormat.CHANNEL_OUT_MONO,
            AudioFormat.ENCODING_PCM_16BIT
        )
        audioTrack = AudioTrack.Builder()
            .setAudioAttributes(
                AudioAttributes.Builder()
                    .setUsage(AudioAttributes.USAGE_MEDIA)
                    .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
                    .build()
            )
            .setAudioFormat(
                AudioFormat.Builder()
                    .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
                    .setSampleRate(24000)
                    .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
                    .build()
            )
            .setBufferSizeInBytes(outBufferSize)
            .setTransferMode(AudioTrack.MODE_STREAM)
            .build()
    }

    private fun startAudioStreaming(expectedGeneration: Long) {
        audioRecord?.startRecording()

        recordThread = Thread {
            val buffer = ByteArray(3200)
            try {
                while (generation.get() == expectedGeneration && started.get()) {
                    val read = audioRecord?.read(buffer, 0, buffer.size) ?: -1
                    if (read > 0) {
                        sendAudio(buffer.copyOf(read))
                    }
                }
            } catch (_: InterruptedException) {}
        }.apply { start() }

        playThread = Thread {
            audioTrack?.play()
            try {
                while (generation.get() == expectedGeneration && started.get()) {
                    Thread.sleep(10)
                }
            } catch (_: InterruptedException) {}
        }.apply { start() }
    }

    private fun playAudioData(data: ByteArray) {
        audioTrack?.write(data, 0, data.size)
        var sum = 0.0
        var i = 0
        while (i < data.size - 1) {
            val sample = (data[i].toInt() and 0xFF or (data[i + 1].toInt() shl 8)).toShort()
            sum += sample.toDouble() * sample.toDouble()
            i += 2
        }
        val rms = kotlin.math.sqrt(sum / (data.size / 2))
        val outputLevel = (rms / Short.MAX_VALUE).coerceIn(0.0, 1.0)
        onLevels?.invoke(0.0, outputLevel)
    }

    private fun stopAudio() {
        recordThread?.interrupt()
        recordThread = null
        playThread?.interrupt()
        playThread = null
        try { audioRecord?.stop() } catch (_: Exception) {}
        try { audioRecord?.release() } catch (_: Exception) {}
        audioRecord = null
        try { audioTrack?.stop() } catch (_: Exception) {}
        try { audioTrack?.release() } catch (_: Exception) {}
        audioTrack = null
    }

    sealed class GeminiTransportException(message: String) : Exception(message) {
        data object microphone : GeminiTransportException("Allow microphone access in Settings.")
        data object connection : GeminiTransportException("Gemini voice connection failed. Check your API key and try again.")
    }
}

File: apps/ios/App/GeminiLiveTransport.swift
import Foundation
import AVFoundation

@MainActor final class GeminiLiveTransport: NSObject, URLSessionWebSocketDelegate {
    var onEvent: (([String: Any]) -> Void)?
    var onLevels: ((Double, Double) -> Void)?
    var onFailure: ((String) -> Void)?

    private var task: URLSessionWebSocketTask?
    private var session: URLSession?
    private var audioEngine: AVAudioEngine?
    private var outputNode: AVAudioPlayerNode?
    private var isStarted = false
    private(set) var isMuted = false
    private var sessionStartTime: TimeInterval = 0
    private var attempt = UUID()
    private var pendingContinuation: CheckedContinuation<Void, Error>?
    private var receiveTask: Task<Void, Never>?

    func connect(api: APIClient, instructions: String, history: [[String: Any]]) async throws {
        disconnect()
        let token = UUID(); attempt = token
        let granted = await AVAudioApplication.requestRecordPermission()
        guard granted else { throw TransportError.microphone }
        try Task.checkCancellation()

        let config = api.config
        let apiKey = config.apiKey
        let model = config.resolvedModel

        let urlString = "wss://generativelanguage.googleapis.com/ws/google.ai.generativelanguage.v1beta.GenerativeService.BidiGenerateContent?key=\(apiKey)"
        guard let url = URL(string: urlString) else { throw TransportError.connection }

        let urlSession = URLSession(configuration: .default, delegate: self, delegateQueue: nil)
        self.session = urlSession
        let wsTask = urlSession.webSocketTask(with: url)
        self.task = wsTask
        wsTask.resume()

        let setupMessage: [String: Any] = [
            "setup": [
                "model": "models/\(model)",
                "responseModalities": ["AUDIO"],
                "speechConfig": [
                    "voiceConfig": [
                        "prebuiltVoiceConfig": ["voiceName": "Kore"]
                    ]
                ],
                "systemInstruction": [
                    "parts": [["text": instructions]]
                ],
                "inputAudioTranscription": [:]
            ]
        ]

        let data = try JSONSerialization.data(withJSONObject: setupMessage)
        guard let jsonString = String(data: data, encoding: .utf8) else { throw TransportError.connection }
        try await wsTask.send(.string(jsonString))

        var receivedSetup = false
        let deadline = Date().addingTimeInterval(15)
        while !receivedSetup && Date() < deadline {
            try Task.checkCancellation()
            guard attempt == token else { throw CancellationError() }
            let message = try await withCheckedThrowingContinuation { (c: CheckedContinuation<URLSessionWebSocketTask.Message, Error>) in
                wsTask.receive { result in c.resume(with: result) }
            }
            if case .string(let text) = message,
               let msgData = text.data(using: .utf8),
               let msg = try? JSONSerialization.jsonObject(with: msgData) as? [String: Any] {
                if msg["setupComplete"] != nil {
                    receivedSetup = true
                    sessionStartTime = Date().timeIntervalSince1970 * 1000
                    onEvent?(["type": "mural.session.created", "session": ["id": "gemini-\(UUID().uuidString)"]])
                    onEvent?(["type": "session.started", "session": ["id": "gemini-\(UUID().uuidString)"]])
                    startAudioStreaming(token: token)
                    startReceiving(token: token)
                }
            }
        }
        guard receivedSetup else { throw TransportError.timeout }
    }

    private func startReceiving(token: UUID) {
        receiveTask?.cancel()
        receiveTask = Task { [weak self] in
            guard let self else { return }
            while !Task.isCancelled {
                guard let wsTask = self.task, self.attempt == token else { return }
                do {
                    let message = try await withCheckedThrowingContinuation { (c: CheckedContinuation<URLSessionWebSocketTask.Message, Error>) in
                        wsTask.receive { result in c.resume(with: result) }
                    }
                    if case .string(let text) = message,
                       let msgData = text.data(using: .utf8),
                       let msg = try? JSONSerialization.jsonObject(with: msgData) as? [String: Any] {
                        await MainActor.run { self.handleServerMessage(msg) }
                    }
                } catch {
                    if !Task.isCancelled {
                        await MainActor.run {
                            self.isStarted = false
                            self.stopAudio()
                            self.onFailure?("Gemini voice connection lost: \(error.localizedDescription)")
                        }
                    }
                    return
                }
            }
        }
    }

    private func handleServerMessage(_ msg: [String: Any]) {
        guard let serverContent = msg["serverContent"] as? [String: Any] else { return }

        if let inputTranscription = serverContent["inputTranscription"] as? [String: Any],
           let text = inputTranscription["text"] as? String, !text.isEmpty {
            let elapsed = Date().timeIntervalSince1970 * 1000 - sessionStartTime
            onEvent?([
                "type": "session.input_transcript.delta",
                "event_id": UUID().uuidString,
                "delta": text,
                "start_ms": max(0, Int(elapsed - 500)),
                "end_ms": Int(elapsed)
            ])
        }

        if let modelTurn = serverContent["modelTurn"] as? [String: Any],
           let parts = modelTurn["parts"] as? [[String: Any]] {
            for part in parts {
                if let inlineData = part["inlineData"] as? [String: Any],
                   let mimeType = inlineData["mimeType"] as? String,
                   let dataStr = inlineData["data"] as? String,
                   mimeType.hasPrefix("audio/"),
                   let audioData = Data(base64Encoded: dataStr) {
                    playAudioData(audioData)
                    continue
                }
                if let text = part["text"] as? String, !text.isEmpty {
                    let elapsed = Date().timeIntervalSince1970 * 1000 - sessionStartTime
                    onEvent?([
                        "type": "session.output_transcript.delta",
                        "event_id": UUID().uuidString,
                        "delta": text,
                        "start_ms": max(0, Int(elapsed - 500)),
                        "end_ms": Int(elapsed)
                    ])
                }
            }
        }

        if let turnComplete = serverContent["turnComplete"] as? Bool, turnComplete {
            let elapsed = Date().timeIntervalSince1970 * 1000 - sessionStartTime
            onEvent?([
                "type": "session.output_transcript.done",
                "event_id": UUID().uuidString,
                "start_ms": max(0, Int(elapsed - 1000)),
                "end_ms": Int(elapsed)
            ])
            onLevels?(0, 0)
        }
    }

    func send(_ event: [String: Any]) -> Bool {
        guard let task, task.state == .running else { return false }
        if let data = try? JSONSerialization.data(withJSONObject: event),
           let json = String(data: data, encoding: .utf8) {
            task.send(.string(json)) { _ in }
            return true
        }
        return false
    }

    func sendText(_ text: String) {
        let msg: [String: Any] = [
            "clientContent": [
                "turns": [["role": "user", "parts": [["text": text]]]],
                "turnComplete": true
            ]
        ]
        _ = send(msg)
    }

    func mute(_ muted: Bool) {
        isMuted = muted
        if muted { onLevels?(0, 0) }
    }

    func close() {
        let closeMsg: [String: Any] = ["clientContent": ["turns": [], "turnComplete": true]]
        _ = send(closeMsg)
        task?.cancel(with: .goingAway, reason: nil)
    }

    func disconnect() {
        attempt = UUID()
        isStarted = false; isMuted = false
        receiveTask?.cancel(); receiveTask = nil
        task?.cancel(with: .goingAway, reason: nil); task = nil
        session?.invalidateAndCancel(); session = nil
        stopAudio()
        onLevels?(0, 0)
    }

    private func startAudioStreaming(token: UUID) {
        let engine = AVAudioEngine()
        let inputNode = engine.inputNode
        let outNode = AVAudioPlayerNode()
        engine.attach(outNode)

        let inputFormat = inputNode.inputFormat(forBus: 0)
        let targetFormat = AVAudioFormat(commonFormat: .pcmFormatInt16, sampleRate: 16000, channels: 1, interleaved: true)!
        let outputFormat = AVAudioFormat(commonFormat: .pcmFormatInt16, sampleRate: 24000, channels: 1, interleaved: true)!
        engine.connect(outNode, to: engine.mainMixerNode, format: outputFormat)

        var converter: AVAudioConverter?
        if inputFormat.sampleRate != 16000 || inputFormat.channelCount != 1 {
            converter = AVAudioConverter(from: inputFormat, to: targetFormat)
        }

        inputNode.installTap(onBus: 0, bufferSize: 3200, format: inputFormat) { [weak self] buffer, _ in
            guard let self, !self.isMuted else { return }
            let sendBuffer: AVAudioPCMBuffer
            if let converter, let converted = converter.convert(buffer) {
                sendBuffer = converted
            } else {
                sendBuffer = buffer
            }
            guard let channelData = sendBuffer.int16ChannelData?[0] else { return }
            let frameLength = Int(sendBuffer.frameLength)
            let pcmData = Data(bytes: channelData, count: frameLength * MemoryLayout<Int16>.size)
            let base64 = pcmData.base64EncodedString()
            let msg: [String: Any] = [
                "realtimeInput": [
                    "audio": ["mimeType": "audio/pcm;rate=16000", "data": base64]
                ]
            ]
            Task { @MainActor [weak self] in self?.send(msg) }
        }

        do {
            try engine.start()
            outNode.play()
            self.audioEngine = engine
            self.outputNode = outNode
            self.isStarted = true
        } catch {
            onFailure?("Failed to start audio: \(error.localizedDescription)")
        }
    }

    private func playAudioData(_ data: Data) {
        guard let outputNode else { return }
        let format = AVAudioFormat(commonFormat: .pcmFormatInt16, sampleRate: 24000, channels: 1, interleaved: true)!
        let frameCount = AVAudioFrameCount(data.count / MemoryLayout<Int16>.size)
        guard let buffer = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: frameCount) else { return }
        buffer.frameLength = frameCount
        data.withUnsafeBytes { ptr in
            guard let baseAddress = ptr.baseAddress else { return }
            buffer.int16ChannelData?[0].update(from: baseAddress.assumingMemoryBound(to: Int16.self), count: Int(frameCount))
        }
        outputNode.scheduleBuffer(buffer, completionHandler: nil)
    }

    private func stopAudio() {
        audioEngine?.stop()
        audioEngine = nil
        outputNode = nil
    }

    nonisolated func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask,
                                didOpenWithProtocol protocol: String?) {}
    nonisolated func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask,
                                didCloseWith closeCode: URLSessionWebSocketTask.CloseCode, reason: Data?) {}
}
